File: frontend/src/preview/actionItems.js
import { SOURCE_RECORDS } from '../lib/actionItems';
const links=['source_type','source_id','review_id','finding_id','risk_id','vendor_id','policy_id','assessment_id'];
export function prepareTask(db,row,previous) {
  if(!['critical','high','medium','low'].includes(row.priority)&&!(previous&&row.priority===previous.priority)) throw new Error('Invalid Action Item priority.');
  if(row.assignee_id&&!db.users.some(u=>u.user_id===row.assignee_id&&(['super_admin','platform_admin'].includes(u.role)||(u.client_ids||[]).includes(row.client_id)))) throw new Error('Assignee must have access to this client.');
  if(previous) {
    if(links.some(k=>(row[k]??null)!==(previous[k]??null))) throw new Error('The originating source and relationships must be retained.');
    if(previous.status==='done'&&row.status!=='done') throw new Error('Completed Action Items remain historical evidence.');
    return;
  }
  if(row.status!=='open') throw new Error('New Action Items start Open.');
  const type=row.source_type||Object.keys(SOURCE_RECORDS).find(t=>row[SOURCE_RECORDS[t][1]])||'manual';
  if(![...Object.keys(SOURCE_RECORDS),'manual','audit'].includes(type)) throw new Error('Invalid source.');
  row.source_type=type;
  if(type==='audit'&&!row.source_id&&!row.assessment_id) {
    if(Object.entries(SOURCE_RECORDS).some(([t,[,key]])=>t!=='audit'&&row[key])) throw new Error('Select the linked record as the source.');
    return;
  }
  if(SOURCE_RECORDS[type]) {
    const [kind,key]=SOURCE_RECORDS[type],id=row.source_id||row[key];
    const source=db[kind].find(r=>r[key]===id&&r.client_id===row.client_id);
    if(!source) throw new Error('Select a source record from this client.');
    if(row[key]&&row[key]!==id) throw new Error('Conflicting source relationship.');
    row[key]=id;row.source_id=id;
    if(row.finding_id) {
      const finding=db.findings.find(f=>f.finding_id===row.finding_id&&f.client_id===row.client_id);
      if(!finding) throw new Error('Finding must belong to this client.');
      for(const field of ['review_id','occurrence_id','vendor_id']) if(finding[field]) {
        if(row[field]&&row[field]!==finding[field]) throw new Error('Conflicting Finding provenance.');
        row[field]=finding[field];
      }
    }
    if(row.review_id) {
      const review=db.reviews.find(r=>r.review_id===row.review_id&&r.client_id===row.client_id);
      if(!review) throw new Error('Review must belong to this client.');
      if(review.vendor_id) {if(row.vendor_id&&row.vendor_id!==review.vendor_id) throw new Error('Conflicting Vendor provenance.');row.vendor_id=review.vendor_id;}
      row.occurrence_id=row.occurrence_id||review.current_occurrence_id||'occ_'+row.review_id;
    }
  } else if(row.source_id||Object.values(SOURCE_RECORDS).some(([,key])=>row[key])) throw new Error('Select the linked record as the source.');
}

File: backend/action_items.py
"""Small Task provenance contract; no new entity or shadow task store."""
from fastapi import HTTPException

SOURCES = {"review": ("reviews", "review_id"), "finding": ("findings", "finding_id"),
           "risk": ("risks", "risk_id"), "vendor": ("vendors", "vendor_id"), "policy": ("policies", "policy_id"), "audit": ("assessments", "assessment_id")}
LINKS = {"source_type", "source_id", "review_id", "finding_id", "risk_id", "vendor_id", "policy_id", "assessment_id"}

async def prepare(db, row, can_access, previous=None):
    cid = row["client_id"]
    if row.get("priority") not in ("critical", "high", "medium", "low") and not (previous and row.get("priority") == previous.get("priority")):
        raise HTTPException(422, "Invalid Action Item priority")
    owner_id = row.get("assignee_id")
    if owner_id:
        owner = await db.users.find_one({"user_id": owner_id}, {"_id": 0})
        if not owner or not can_access(owner, cid):
            raise HTTPException(422, "Assignee must have access to this client")
    if previous:
        if any(row.get(k) != previous.get(k) for k in LINKS):
            raise HTTPException(422, "The originating source and relationships must be retained")
        if previous.get("status") == "done" and row.get("status") != "done":
            raise HTTPException(409, "Completed Action Items remain historical evidence")
        return row
    if row.get("status", "open") != "open":
        raise HTTPException(422, "New Action Items start Open")
    source_type = row.get("source_type") or next((t for t, (_, key) in SOURCES.items() if row.get(key)), "manual")
    if source_type not in {*SOURCES, "manual", "audit"}:
        raise HTTPException(422, "Invalid source")
    row["source_type"] = source_type
    if source_type == "audit" and not row.get("source_id") and not row.get("assessment_id"):
        if any(row.get(key) for typ, (_, key) in SOURCES.items() if typ != "audit"):
            raise HTTPException(422, "Select the linked record as the source")
        return row
    if source_type in SOURCES:
        kind, key = SOURCES[source_type]
        ident = row.get("source_id") or row.get(key)
        source = await db[kind].find_one({key: ident, "client_id": cid}, {"_id": 0}) if ident else None
        if not source:
            raise HTTPException(422, "Select a source record from this client")
        if row.get(key) and row[key] != ident:
            raise HTTPException(422, "Conflicting source relationship")
        row.update({key: ident, "source_id": ident})
        if row.get("finding_id"):
            source_finding = await db.findings.find_one({"finding_id": row["finding_id"], "client_id": cid}, {"_id": 0})
            if not source_finding:
                raise HTTPException(422, "Finding must belong to this client")
            for field in ("review_id", "occurrence_id", "vendor_id"):
                if source_finding.get(field):
                    if row.get(field) and row[field] != source_finding[field]:
                        raise HTTPException(422, "Conflicting Finding provenance")
                    row[field] = source_finding[field]
        if row.get("review_id"):
            review = await db.reviews.find_one({"review_id": row["review_id"], "client_id": cid}, {"_id": 0})
            if not review:
                raise HTTPException(422, "Review must belong to this client")
            if review.get("vendor_id"):
                if row.get("vendor_id") and row["vendor_id"] != review["vendor_id"]:
                    raise HTTPException(422, "Conflicting Vendor provenance")
                row["vendor_id"] = review["vendor_id"]
            # A Finding's historical execution wins over the Review's current execution.
            row["occurrence_id"] = row.get("occurrence_id") or review.get("current_occurrence_id") or "occ_" + review["review_id"]
    elif row.get("source_id") or any(row.get(key) for _,key in SOURCES.values()):
        raise HTTPException(422, "Select the linked record as the source")
    return row
